// pnpm-shrinkwrap/lib/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
__export(require("./prune"));
__export(require("./read"));
const existsWanted_1 = require("./existsWanted");
exports.existsWanted = existsWanted_1.default;
const nameVerFromPkgSnapshot_1 = require("./nameVerFromPkgSnapshot");
exports.nameVerFromPkgSnapshot = nameVerFromPkgSnapshot_1.default;
const pkgSnapshotToResolution_1 = require("./pkgSnapshotToResolution");
exports.pkgSnapshotToResolution = pkgSnapshotToResolution_1.default;
const satisfiesPackageJson_1 = require("./satisfiesPackageJson");
exports.satisfiesPackageJson = satisfiesPackageJson_1.default;
const write_1 = require("./write");
exports.write = write_1.default;
exports.writeCurrentOnly = write_1.writeCurrentOnly;
exports.writeWantedOnly = write_1.writeWantedOnly;
// for backward compatibility
const dependency_path_1 = require("dependency-path");
exports.getPkgShortId = dependency_path_1.refToRelative;
