diff --git a/.bazelversion b/.bazelversion
index 48f7a71df4..ae6dd4e203 100644
--- a/.bazelversion
+++ b/.bazelversion
@@ -1 +1 @@
-0.28.1
+0.29.0
diff --git a/WORKSPACE b/WORKSPACE
index 699a69adda..ad9c3b3051 100644
--- a/WORKSPACE
+++ b/WORKSPACE
@@ -10,10 +10,10 @@ load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
 
 http_archive(
     name = "io_bazel_rules_closure",
-    sha256 = "b6936ecc0b5a1ef616b9d7e76694d414aa5605265c11322257a610fb256b1bf7",
-    strip_prefix = "rules_closure-7434c41542ca9e1b05166d897b90073d1b8b2cf8",
+    sha256 = "2e95ba060acd74f3662547a38814ffff60317be047b7168d25498aea52f3e732",
+    strip_prefix = "rules_closure-b3d4ec3879620edcadd3422b161cebb37c59b6c5",
     urls = [
-        "https://github.com/bazelbuild/rules_closure/archive/7434c41542ca9e1b05166d897b90073d1b8b2cf8.tar.gz",
+        "https://github.com/bazelbuild/rules_closure/archive/b3d4ec3879620edcadd3422b161cebb37c59b6c5.tar.gz",
     ],
 )
 
@@ -36,8 +36,8 @@ selenium_register_dotnet()
 
 http_archive(
     name = "build_bazel_rules_nodejs",
-    sha256 = "6d4edbf28ff6720aedf5f97f9b9a7679401bf7fca9d14a0fff80f644a99992b4",
-    urls = ["https://github.com/bazelbuild/rules_nodejs/releases/download/0.32.2/rules_nodejs-0.32.2.tar.gz"],
+    sha256 = "da72ea53fa1cb8ab5ef7781ba06b97259b7d579a431ce480476266bc81bdf21d",
+    urls = ["https://github.com/bazelbuild/rules_nodejs/releases/download/0.36.2/rules_nodejs-0.36.2.tar.gz"],
 )
 
 load("@build_bazel_rules_nodejs//:defs.bzl", "npm_install")
@@ -55,10 +55,10 @@ install_bazel_dependencies()
 http_archive(
     name = "rules_python",
     urls = [
-      "https://github.com/bazelbuild/rules_python/archive/9d68f24659e8ce8b736590ba1e4418af06ec2552.zip"
+      "https://github.com/bazelbuild/rules_python/archive/e0644961d74b9bbb8a975a01bebb045abfd5d1bd.zip"
     ],
-    strip_prefix = "rules_python-9d68f24659e8ce8b736590ba1e4418af06ec2552",
-    sha256 = "f7402f11691d657161f871e11968a984e5b48b023321935f5a55d7e56cf4758a",
+    strip_prefix = "rules_python-e0644961d74b9bbb8a975a01bebb045abfd5d1bd",
+    sha256 = "b556b165ea1311bf68b6b0bc86d95e5cfca2e839aa6fbd232781bb3930f3d392",
 )
 
 # This call should always be present.
