diff --git a/rb/spec/integration/selenium/webdriver/edge/service_spec.rb b/rb/spec/integration/selenium/webdriver/edge/service_spec.rb
index 40fd531..ad68ce5 100644
--- a/rb/spec/integration/selenium/webdriver/edge/service_spec.rb
+++ b/rb/spec/integration/selenium/webdriver/edge/service_spec.rb
@@ -19,7 +19,7 @@
 
 require_relative '../spec_helper'
 
-      module Selenium
+module Selenium
   module WebDriver
     module Edge
       describe Service,
