--- a/MODULE.bazel
+++ b/MODULE.bazel
@@ -7,6 +7,7 @@

 # Lower-bounds (minimum) versions for direct runtime dependencies
 bazel_dep(name = "aspect_bazel_lib", version = "2.7.7")
+bazel_dep(name = "bazel_lib", version = "3.0.0")
 bazel_dep(name = "aspect_rules_js", version = "2.0.0")
 bazel_dep(name = "aspect_tools_telemetry", version = "0.2.8")
 bazel_dep(name = "bazel_skylib", version = "1.5.0")
