diff --git a/ruby/private/gem_push.bzl b/ruby/private/gem_push.bzl
index b174e55..a2eee96 100644
--- a/ruby/private/gem_push.bzl
+++ b/ruby/private/gem_push.bzl
@@ -10,17 +10,20 @@ def _rb_gem_push_impl(ctx):
     if ctx.attr.ruby != None:
         ruby_toolchain = ctx.attr.ruby[platform_common.ToolchainInfo]
     srcs = [ctx.file.gem]
-    tools = [ruby_toolchain.gem]
+    tools = list(ruby_toolchain.files)
+    java_bin = ""
 
     if ruby_toolchain.version.startswith("jruby"):
         env["JAVA_HOME"] = java_toolchain.java_runtime.java_home
         tools.extend(java_toolchain.java_runtime.files.to_list())
+        java_bin = java_toolchain.java_runtime.java_executable_runfiles_path[3:]
 
     script = generate_rb_binary_script(
         ctx,
         binary = ruby_toolchain.gem,
         bundler = False,
         args = ["push", ctx.file.gem.short_path],
+        java_bin = java_bin,
     )
 
     runfiles = ctx.runfiles(srcs + tools)
